app/src/androidTest/java/org/simple/clinic/pin/BruteForceProtectionAndroidTest.kt

```diff
@@ -0,0 +1,115 @@
+package org.simple.clinic.pin
+
+import android.support.test.annotation.UiThreadTest
+import android.support.test.runner.AndroidJUnit4
+import com.f2prateek.rx.preferences2.Preference
+import com.google.common.truth.Truth.assertThat
+import io.reactivex.Single
+import io.reactivex.plugins.RxJavaPlugins
+import io.reactivex.schedulers.TestScheduler
+import org.junit.After
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.simple.clinic.TestClinicApp
+import org.simple.clinic.pin.BruteForceProtection.ProtectedState.Allowed
+import org.simple.clinic.pin.BruteForceProtection.ProtectedState.Blocked
+import org.simple.clinic.util.Optional
+import org.simple.clinic.util.TestClock
+import org.threeten.bp.Clock
+import org.threeten.bp.Duration
+import org.threeten.bp.Instant
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+import javax.inject.Named
+
+@RunWith(AndroidJUnit4::class)
+class BruteForceProtectionAndroidTest {
+
+  @Inject
+  lateinit var clock: Clock
+
+  @Inject
+  lateinit var bruteForceProtection: BruteForceProtection
+
+  @Inject
+  lateinit var configProvider: Single<BruteForceProtectionConfig>
+
+  @Inject
+  @field:Named("pin_failed_auth_count")
+  lateinit var failedAttemptsCount: Preference<Int>
+
+  @Inject
+  @field:Named("pin_failed_auth_limit_reached_at")
+  lateinit var blockedAt: Preference<Optional<Instant>>
+
+  private val testClock
+    get() = clock as TestClock
+
+  private val config
+    get() = configProvider.blockingGet()
+
+  private val testScheduler = TestScheduler()
+
+  @Before
+  fun setup() {
+    TestClinicApp.appComponent().inject(this)
+
+    RxJavaPlugins.setComputationSchedulerHandler { testScheduler }
+
+    failedAttemptsCount.delete()
+    blockedAt.delete()
+  }
+
+  @Test
+  fun when_listening_to_protected_state_changes_then_correct_state_changes_should_be_emitted() {
+    val initialState = bruteForceProtection.protectedStateChanges().blockingFirst()
+    assertThat(initialState).isEqualTo(Allowed)
```

Comment:
> It does not verify that the protectedStateChanges emission when the failed attempt count is more than one and less than the max number of attempts is Allowed and not Blocked. We should be testing the state between each attempt as well.

Valid point. I'll do this.

> It does not verify that the time at which the blockedTill Instant is calculated is from the moment when the last attempt fails.

That… is a good point. I thought I was already doing that because I'm calculating the expected `blockedTill` time, but the time remains constant through the test. I should advance the time after each failed attempt.

Reply:
> I'd suggest not using a repeat operator and overriding the config to a value set in the test so you can manually increment the failed attempt counter and verify at each step.

I didn't fully understand this. Will ask you in person.

Reply:
@vinaysshenoy is it okay if I make these changes in my next PR? The feature is behind a flag anyway.

Reply:
Yup.